From bb0381738eeab6e82fc3f20677f74194e4a2bff9 Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Mon, 19 May 2008 20:30:55 +0000
Subject: [PATCH] CONFIGURATION-328: Fixed a bug in XMLConfiguration.addNodes()
 that prevented attribute nodes to be added (bug fix from trunk ported to this
 branch)

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/branches/configuration2_experimental@657962 13f79535-47bb-0310-9956-ffa450edef68
---
 .../configuration2/XMLConfiguration.java      |  1 +
 .../configuration2/TestXMLConfiguration.java  | 21 +++++++++++++++++++
 xdocs/changes.xml                             |  4 ++++
 3 files changed, 26 insertions(+)

diff --git a/src/main/java/org/apache/commons/configuration2/XMLConfiguration.java b/src/main/java/org/apache/commons/configuration2/XMLConfiguration.java
index 2b181d105f..3b281e4f20 100644
--- a/src/main/java/org/apache/commons/configuration2/XMLConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration2/XMLConfiguration.java
@@ -878,6 +878,7 @@ private XMLNode convertToXMLNode(ConfigurationNode node)
 
         XMLNode xmlNode = (XMLNode) createNode(null, node.getName());
         xmlNode.setValue(node.getValue());
+        xmlNode.setAttribute(node.isAttribute());
         for (ConfigurationNode child : node.getChildren())
         {
             xmlNode.addChild(convertToXMLNode(child));
diff --git a/src/test/java/org/apache/commons/configuration2/TestXMLConfiguration.java b/src/test/java/org/apache/commons/configuration2/TestXMLConfiguration.java
index a85aa28a1e..eb4e43627c 100644
--- a/src/test/java/org/apache/commons/configuration2/TestXMLConfiguration.java
+++ b/src/test/java/org/apache/commons/configuration2/TestXMLConfiguration.java
@@ -1364,6 +1364,27 @@ public void testCopyRootNameNoDocument() throws ConfigurationException
                 .getRootElementName());
     }
 
+    /**
+     * Tests adding an attribute node using the addNodes() method.
+     */
+    public void testAddNodesAttributeNode()
+    {
+        conf.addProperty("testAddNodes.property[@name]", "prop1");
+        conf.addProperty("testAddNodes.property(0).value", "value1");
+        conf.addProperty("testAddNodes.property(-1)[@name]", "prop2");
+        conf.addProperty("testAddNodes.property(1).value", "value2");
+        Collection<ConfigurationNode> nodes = new ArrayList<ConfigurationNode>();
+        nodes.add(new DefaultConfigurationNode("property"));
+        conf.addNodes("testAddNodes", nodes);
+        nodes.clear();
+        ConfigurationNode nd = new DefaultConfigurationNode("name", "prop3");
+        nd.setAttribute(true);
+        nodes.add(nd);
+        conf.addNodes("testAddNodes.property(2)", nodes);
+        assertEquals("Attribute not added", "prop3", conf
+                .getString("testAddNodes.property(2)[@name]"));
+    }
+
     /**
      * Prepares a configuration object for testing a reload operation.
      *
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index 115ac254f9..f81db5cb2d 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -80,6 +80,10 @@
     </release>
 
     <release version="1.6" date="in SVN" description="">
+      <action dev="oheger" type="fix" issue="CONFIGURATION-328">
+        A bug in XMLConfiguration.addNodes() made it impossible to add
+        attribute nodes using this method. This has been fixed.
+      </action>
       <action dev="ebourg" type="fix" issue="CONFIGURATION-322">
         ConfigurationDynaBean now works properly with indexed properties
         stored internally in the underlying configuration as arrays.
